sql: disallow adding column as primary key

jasonmchan opened the pull request:
Previously, the behavior of ALTER TABLE ... ADD COLUMN ... PRIMARY KEY
was undefined. With the legacy schema changer, this statement would
appear to succeed, but it would break the new schema changer. This
commit changes this by returning an explicit error.

In the future, we should support this statement if the table's primary
key was originally the default rowid primary key (see #82735).

Fixes #82489

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/82489)
<!-- Reviewable:end -->

Commit:
sql: disallow adding column as primary key

Previously, the behavior of ALTER TABLE ... ADD COLUMN ... PRIMARY KEY
was undefined. With the legacy schema changer, this statement would
appear to succeed, but it would break the new schema changer. This
commit changes this by returning an explicit error.

In the future, we should support this statement if the table's primary
key was originally the default rowid primary key (see #82735).

Fixes #81449

Release note: None

chengxiong-ruan reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82489#-:-N4CsQVr7I2s878RD3hA:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @chengxiong-ruan and @jasonmchan)

___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_add_column.go` line 81 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/82489#-N4CrFsx2pX2KsUlAQej:-N4CrFsx2pX2KsUlAQek:bly5mbh) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ce9b4f59199c4c7b03a68629075733431cc22b1/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_add_column.go#L81)):*
> ```Go
> 		))
> 	}
> 	if d.PrimaryKey.IsPrimaryKey {
> ```

Could you add a TODO and file an issue that we'd need to support this in new schema changer in the future?
I understand that there's always a primary key in tables and we can simply block `ADD COLUMN... PRIMARY KEY` like this. But I think we should still consider it legit for user to do `DROP CONSTRAINT` to drop the primary key following with a `ADD COLUMN... PRIMARY KEY` in a same transaction.
But.. in general people won't be this crazy and there's always a workaround to add the column first then alter primary key. Still.. keep an issue tracking this won't hurt :)

Another thing that, could add a logic to check the target status of primary key is public here as well? I understand it's not necessary, but I think it'd make more sense together with the TODO I requested above.


<!-- Sent from Reviewable.io -->

jasonmchan reviewed (commented):
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82489#-:-N4De8KVAL0tpwrZwd-t:b-814g8k)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @chengxiong-ruan)

___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_add_column.go` line 81 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/82489#-N4CrFsx2pX2KsUlAQej:-N4DdS7X68RYwzi6w2gP:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ce9b4f59199c4c7b03a68629075733431cc22b1/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_add_column.go#L81)):*
<details><summary><i>Previously, chengxiong-ruan (Chengxiong Ruan) wrote…</i></summary><blockquote>

Could you add a TODO and file an issue that we'd need to support this in new schema changer in the future?
I understand that there's always a primary key in tables and we can simply block `ADD COLUMN... PRIMARY KEY` like this. But I think we should still consider it legit for user to do `DROP CONSTRAINT` to drop the primary key following with a `ADD COLUMN... PRIMARY KEY` in a same transaction.
But.. in general people won't be this crazy and there's always a workaround to add the column first then alter primary key. Still.. keep an issue tracking this won't hurt :)

Another thing that, could add a logic to check the target status of primary key is public here as well? I understand it's not necessary, but I think it'd make more sense together with the TODO I requested above.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

chengxiong-ruan reviewed (approved):
> Reviewed 2 of 3 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82489)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @chengxiong-ruan)



<!-- Sent from Reviewable.io -->

jasonmchan commented:
bors r+

ajwerner commented:
bors r+
